plugin/highlights.py

```diff
@@ -0,0 +1,102 @@
+import sublime_plugin
+
+from .core.configurations import is_supported_syntax
+from .core.protocol import Request, Range, DocumentHighlightKind
+from .core.clients import client_for_view
+from .core.documents import get_document_position
+from .core.settings import settings
+
+import sublime  # only for typing
+try:
+    from typing import List, Dict
+    assert List and Dict
+except ImportError:
+    pass
+
+
+_kind2name = {
+    DocumentHighlightKind.Unknown: "unknown",
+    DocumentHighlightKind.Text: "text",
+    DocumentHighlightKind.Read: "read",
+    DocumentHighlightKind.Write: "write"
+}
+
+
+class DocumentHighlightListener(sublime_plugin.ViewEventListener):
+
+    @classmethod
+    def is_applicable(cls, settings):
+        syntax = settings.get('syntax')
+        return syntax and is_supported_syntax(syntax)
+
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+        self._initialized = False
+        self._enabled = False
+        self._version = 0
+
+    def on_selection_modified_async(self) -> None:
+        if not self._initialized:
+            self._initialize()
+        if self._enabled:
+            self._clear_regions()
+            self._queue()
+
+    def _queue(self) -> None:
+        self._version += 1
+        current_version = self._version
+        sublime.set_timeout_async(lambda: self._purge(current_version), 500)
+
+    def _purge(self, this_version: int) -> None:
+        if this_version == self._version:
+            self._on_document_highlight()
+
+    def _clear_regions(self) -> None:
+        for kind in settings.document_highlight_scopes.keys():
+            self.view.erase_regions("LspDocumentHighlightListener_{}".format(kind))
+
+    def _on_document_highlight(self) -> None:
+        self._clear_regions()
+        if len(self.view.sel()) != 1:
+            return
+        point = self.view.sel()[0].begin()
+        if self.view.match_selector(point, "comment"):
+            # We're inside a comment, go home.
+            return
+        client = client_for_view(self.view)
+        if client:
+            params = get_document_position(self.view, point)
+            if params:
+                request = Request.documentHighlight(params)
+                client.send_request(request, self._handle_response)
+
+    def _handle_response(self, response: list) -> None:
+        if not response:
+            return
+        kind2regions = {}  # type: Dict[str, List[sublime.Region]]
+        for kind in range(0, DocumentHighlightKind.Count):
+            kind2regions[_kind2name[kind]] = []
+        for highlight in response:
+            if highlight:
```

Comment:
Is this check needed?

Reply:
I think strictly no. I've removed the check in cc3f1a7.